### PR TITLE
[WIP] Create calculator registry implementation

### PR DESCRIPTION
**Pull request by sagikazarmark**

### CODE REVIEW

**Comment by frederikbosch** [comment]
Looks good!

**Comment by frederikbosch** [comment]
@sagikazarmark I was thinking. We could decide to not make the `Registry` a static class. And set it as a static property of Money, like this. You would agree that we must avoid static classes. However, I also see that it involves some additional code. What is the best solution here, and why?

```php
// create a register
$registry = new Registry([
  new BcMathCalculator(),
  new GmpCalculator(),
  new PhpCalculator(),
]);

// Money class
final class Money {
  private function getCalculatorRegister() {
    if (!self::$calculatorRegister) {
      self::$calculatorRegister = Register::fromDefaultExtensions();
    }
    return self::$calculatorRegister;
  }
  private function getCalculator() {
    return self::getCalculatorRegister()->getCalculator();
  }
  /** @deprecated **/
  public static function registerCalculator($calculator) {
    self::getCalculatorRegister()->register($calculator);
  }
  public static function setCalculatorRegistry($registry) {
    self::$calculatorRegister = $register;
  }
}
```

**Comment by sagikazarmark** [comment]
Hm, yeah, great idea, would definitely make testing easier. However, there has to be some sort of static initialization, preferably outside of money, because the whole point of this separation is that other components can access the same calculator context.

Will work on this, let's see what I can come up with.

**Comment by sagikazarmark** [comment]
@frederikbosch okay, so I made the registry a singleton, added a few features and reorganized internals.

But somehow it still doesn't feel right this way. Could use some fresh eyes. 👁

File: src/Calculator/Registry.php [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+        $this->calculators[$calculator] = false;
+
+        $this->resetCalculator();
+    }
+
+    /**
+     * Disables arbitrary precision calculators.
+     */
+    public function disableArbitraryPrecisionCalculators()
```
Comment (frederikbosch) [inline src/Calculator/Registry.php]: Rather would say you want something like `Register::phpOnly()`.

File: src/Calculator/Registry.php [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+        $this->resetCalculator();
+    }
+
+    /**
+     * Enables a calculator.
+     *
+     * @param string $calculator
+     */
+    public function enableCalculator($calculator)
```
Comment (frederikbosch) [inline src/Calculator/Registry.php]: I would remove the suffix `Calculator`.

File: src/Calculator/Registry.php [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+    /**
+     * @var Registry
+     */
+    private static $instance;
+
+    /**
+     * @param string $calculator
+     */
+    public function registerCalculator($calculator)
```
Comment (frederikbosch) [inline src/Calculator/Registry.php]: I would remove the suffix `Calculator`.

File: src/Calculator/Registry.php [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
+        $this->resetCalculator();
+    }
+
+    /**
+     * Disables a calculator.
+     *
+     * @param string $calculator
+     */
+    public function disableCalculator($calculator)
```
Comment (frederikbosch) [inline src/Calculator/Registry.php]: I would remove the suffix `Calculator`.

File: src/Calculator/Registry.php [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+    {
+        $this->disableCalculator(BcMathCalculator::class);
+        $this->disableCalculator(GmpCalculator::class);
+    }
+
+    /**
+     * Resets the already resolved calculator instance.
+     * Used when the registry is modified in any way (eg. registering a new calculator).
+     */
+    private function resetCalculator()
```
Comment (frederikbosch) [inline src/Calculator/Registry.php]: I would remove the suffix `Calculator`.

File: src/Calculator/Registry.php [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+    {
+        $this->calculator = null;
+    }
+
+    /**
+     * @return Calculator
+     *
+     * @throws \RuntimeException If cannot find calculator for money calculations
+     */
+    private function initializeCalculator()
```
Comment (frederikbosch) [inline src/Calculator/Registry.php]: I would remove the suffix `Calculator`.

File: src/Calculator/Registry.php [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+            }
+        }
+
+        throw new \RuntimeException('Cannot find calculator for money calculations');
+    }
+
+    /**
+     * @return Calculator
+     */
+    public function getCalculator()
```
Comment (frederikbosch) [inline src/Calculator/Registry.php]: I would remove the suffix `Calculator`.

**Comment by frederikbosch** [comment]
Superseded by #646

**Pull request closed without merging**